### ISSUE TITLE
Define and adapt the QA team repositories

### ISSUE THREAD

**Issue by damarisg** [wazuh/wazuh-qa#5475]
## Description

To improve the structure and maintainability of the `wazuh-qa` and `wazuh-jenkins` repositories, we need to define and adapt the current design. This will help organize tests and improve practices.

## Functional requirements

- The structure should be organized according to the types of tests. 
- The structure must adapt to continuous growth, allowing the incorporation of new features.
- The structure must maintain consistency to ensure uniformity.
- Each section should include appropriate documentation to assist quality control tasks.
- The structure should make automated tests easy to implement and run.
- The new design must integrate with supported build systems and pipelines.

## Tasks

- [x] [Create a structure](https://github.com/wazuh/wazuh-qa/issues/5475#issuecomment-2239970249). (all)
- [x] Update the repository structure as approved. (@QU3B1M)
- [x] Document the justification of the proposed structure. (@QU3B1M)
- [x] Document the processes, good practices, and sections defined in this proposal. (@QU3B1M)
- [x] Research the list of issues to remove from https://github.com/wazuh/wazuh-qa. (@damarisg )
- [x] Research the list of issues to remove from https://github.com/wazuh/wazuh-jenkins.  (@damarisg )

---

## Wazuh QA automation repository

This repository will store the Wazuh QA framework, the different test suites, and the QA Jenkins environment code. In the initial version, only some basic tools of the different repositories will be developed/migrated, with the Deployability tests suite, and the new Jenkins environment with the approach of Everything-as-a-code.  

### Repository structure

After analyzing the current repository's structure, and the different proposals, we can conclude that our new structure must follow the next key points:
- **It must be scalable**.
- Tooling and test suite must be easy to understand and maintain.
- Tools and utils must have unit tests.
- Test types must be well-defined and separated.
- Tests-specific helpers or tools must be in the same directory as the tests.

Taking into account the previous points, we propose the following structure:

```shellsession
.
├── README.md
├── tests
│   ├── test_functional
│   │   ├── test_end_to_end
│   │   │   ├── helpers
│   │   └── test_system
│   │       ├── helpers
│   │       └── test_deployability
│   └── test_non_functional
│       ├── test_performance
│       └── test_stress
│           └── test_sca
│               └── helpers
└── tools
    ├── src
    │   ├── generic
    │   │   ├── logger
    │   │   └── monitoring
    │   ├── jenkins
    │   │   ├── bootstrap
    │   │   ├── jenkins-ui
    │   │   └── procedures
    │   ├── jobflow
    │   ├── provision
    │   ├── publishing
    │   └── wazuh_qa_framework              # Name to be discussed
    └── tests
        ├── generic
        ├── jenkins
        ├── jobflow
        ├── provision
        ├── publishing
        └── wazuh_qa_framework
```

<details><summary>Other option</summary>

```shellsession
.
├── README.md
├── tests
│   ├── test_functional
│   │   └── test_system
│   │       └── test_deployability
│   └── test_non_functional
│       └── test_stress
│           └── test_sca
└── tools
    ├── generic
    │   ├── logger
    │   └── monitoring
    ├── jenkins
    │   ├── src
    │   │   ├── bootstrap
    │   │   ├── jenkins-ui
    │   │   └── procedures
    │   └── tests
    ├── jobflow
    │   ├── src
    │   └── test
    ├── provision
    │   ├── src
    │   └── test
    └── wazuh_qa_framework
        ├── src
        │   ├── api_client
        │   ├── constants
        │   └── utils
        └── tests
```
</details>

### Proposed structure documentation and justification

Currently, we are facing two issues. **First**, our main `wazuh-qa` repository has a lot of useful code, but as its structure is not well defined, it is difficult to maintain and scale, the test types are not well defined, and the tools and utils are all part of a big framework that does everything (its scope is not well defined). **Second**, we have other repositories with useful code, but they are not part of the main repository, so we have content spread across different repositories making it hard to maintain.  
So, the idea is to avoid this situation by creating a new structure that allows us to scale and maintain the code easily, having a well-defined scope for each tool and test type. It also includes code from the other repositories owned by QA

To achieve this goal, the proposed structure is divided into two main directories: `tests/` and `tools/` which at the same time are divided into subdirectories with more subdirectories, going from a generic scope to a more specific one.

The `tests/` directory is divided into the main tests categories _functional_ and _non-functional_, from that point the categories can be divided into more specific test types, such as _end-to-end_, and _system_ for the _functional tests_, and _performance_ and _stress_ for the _non-functional tests_. Each of these directories could contain helpers or [fixtures](https://docs.pytest.org/en/7.1.x/reference/fixtures.html) used only by the tests in that scope and its subdirectories.

For the `tools/` directory, the structure is different, we don't have 'categories', in this directory resides all the tools that we use in our QA process, such as _Jenkins_, _JobFlow_, _Provision_, and _wazuh_qa_framework_, alongside all the unit tests for each tool. Its basic structure is divided into `src/` and `tests/`, where the `src/` directory contains the source code of the tools, and the `tests/` directory contains the unit tests for the tools.
Another thing to consider is that the `wazuh_qa_framework` is defined only to be a 'wazuh components handler' tool, this way we can have a well-defined scope for this tool, and avoid making it 'infinite' as it is now.

#### Initial migration process

In this first approach, we will only migrate the newly developed tools and tests by the QA team, which are DTT1 and the new Jenkins environment. This migration flow consists of the following simple steps: 
    1. Review the current code
    2. Analyze and propose an adaptation for the new structure
    3. Migrate the code with the pertinent modifications 

As the code is recently developed, we don't expect to have many issues during the migration process, but if we find any, we must document them and propose a solution.

#### Tests development process

The development of new tests must follow a well-defined process to ensure that the tests are well-developed and maintained. A proposal for this process is defined as follows:
    1. Analyze and define the scope and type of the test scenario. (could use a guide like [this](https://www.browserstack.com/guide/types-of-testing))
    2. Define a proposal set of Test Cases
    3. Develop the approved test cases with the required helpers or fixtures

**Other Recommendations**
- The tests must be written using the Gherkin syntax. ([pytest-bdd](https://pypi.org/project/pytest-bdd/))
- The E2E tests must be developed using the Page Object Model pattern.
- Use the [pytest](https://docs.pytest.org/en/) framework to develop the tests.
- Use [playwright](https://playwright.dev/python/docs/intro) to interact with the browser in the E2E tests.

#### Tools development process

The tools development process must also follow a well-defined process to ensure its correct development. A proposal for this process is defined as follows:
    1. Analyze the requested feature or tool
    2. Define the structure of the tool and technologies to use
    3. Develop the tool with the required unit tests

#### Good practices

- **Gherkin syntax**: All the tests must be written using the Gherkin syntax. ([pytest-bdd](https://pypi.org/project/pytest-bdd/))
- **Code review**: All the code must be reviewed by at least one other team member before merging it into the main branch.
- **Unit tests**: All the tools and tests must have unit tests.
- **Documentation**: All the code must be documented, and the documentation must be updated with the code.
- **Code style**: All the code must follow the same code style, and the code style must be defined in the documentation. (In the case of Python, we must follow [PEP8](https://pep8.org/))
- **Continuous Integration**: All the code must pass through a series of defined GitHub Actions to ensure:
    - The code style is correct
    - The unit tests pass
    - The code is well documented
    - The code and documentation are typos-free

#### Useful sources

Development
- [PEP8](https://pep8.org/)
- [PEP257](https://peps.python.org/pep-0257/)
- [PEP20](https://peps.python.org/pep-0020/)
- [SOLID principles](https://realpython.com/solid-principles-python/)
- [Pythonic code](https://builtin.com/data-science/pythonic#:~:text=What%20Is%20Pythonic%20Code,and%20not%20deciphering%20existing%20code.)
- [Python Best practices](https://realpython.com/tutorials/best-practices/

QA Automation
- [Software Testing Types: The Ultimate Guide](https://www.browserstack.com/guide/types-of-testing)
- [Software Testing Life Cycle (STLC)](https://www.guru99.com/software-testing-life-cycle.html)
- [Gherkin syntax](https://cucumber.io/docs/gherkin/)
- [Page Object Model](https://playwright.dev/docs/pom)
- [Pytest documentation](https://docs.pytest.org/en/)
- [Pytest-bdd documentation](https://pypi.org/project/pytest-bdd/)

**Comment by damarisg** [comment]
## Migration

- [wazuh-tools](https://github.com/wazuh/wazuh-tools/): 
   - [footprint/sca](https://github.com/wazuh/wazuh-tools/tree/master/footprint/sca) 

## Structure

```shellsession
[wazuh-qa]
|
|─tests
|    ├ non_functional
|          ├ footprint
|              └── sca
|    └ functional
|          ├ component
|          ├ integration
|          └ e2e

```

**Comment by QU3B1M** [comment]
## Update report

Working with @pcapellan in parallel to define possible structures and technologies for the new QA repository that will store the QA framework, Jenkins, and test code. 

Two boilerplate versions were currently presented in a quick sync through the Slack chat, but they are still Work in Progress.

**Comment by damarisg** [comment]
## 15/07/2024 

- Work on `Research the list of issues to remove from wazuh-qa.`

**Comment by jnasselle** [comment]
# Review

Great job @wazuh/devel-qa-div3 ! some comments based on what we've spoken:

- GHA directory to enforce whatever we need and execute UT
- tools should contain self-contained structure for better comprehension i.e rather than having src/generic/jobflow and tests/jobflow, i suggest tools/jobflow/src and /tools/jobflow/tests
- Tests development process, Other recommendations, Tools development process and Good practices are critical information in order to improve Wazuh quality standards, and should be taken into account, but is out of scope of the current issue goal.
- tests/test_functional and tests/test_no_functional could be simplified by removing test prefix

**Comment by pcapellan** [comment]
I think it's important to define if we are going to write the tests with the gherkin syntax and if we are going to use the page object model pattern for e2e test cases. If this were the case I would suggest, for the testing section, a structure like this:

```shellsession
qa-wazuh/
│
├── features/
│   ├── feature_files/		# contains gherkin files and test steps
│   │   ├── logcollector.feature
│   │   └── syscheck.feature
│   └── steps/			# contains the python files that implement the steps defined in feature_files
│       ├── __init__.py
│       ├── logcollector_steps.py
│       └── syscheck_steps.py
│
├── pages/			# applies for e2e tests
│   ├── __init__.py
│   ├── events_page.py
│   └── dashboard_page.py
│   └── inventory_page.py
│
│
└── tests/
    ├── __init__.py
    ├── test_cases/		#  contains the test files written with pytest
    │   ├── __init__.py
    │   ├── test_logcollector.py
    │   └── test_syscheck.py
    │
    └── resources/		# contains resources for testing: data generation, configurations, etc.
 
```

**Comment by QU3B1M** [comment]
## Structure Update

Considering the previous recommendations, a possible structure could look something like this

```yaml
.
├── .github
│   └── workflows
│       ├── run_unitests.yaml
│       └── tools_coverage.yaml
├── README.md
├── tests
│   ├── test_functional
│   │   ├── test_e2e
│   │   │   ├── features
│   │   │   ├── page_objects
│   │   │   └── tests               # Or steps_definitions/
│   │   └── test_system
│   │       └── test_deployability
│   │           ├── features
│   │           └── tests           # Or steps_definitions/
│   └── test_non_functional
│       └── test_stress
│           ├── features
│           └── tests               # Or steps_definitions/
└── tools
    ├── generic
    │   ├── src
    │   │   ├── logger
    │   │   └── monitors
    │   └── tests
    │       ├── logger
    │       └── monitors
    ├── jenkins
    │   ├── src
    │   └── tests
    ├── jobflow
    │   ├── src
    │   └── test
    ├── provision
    │   ├── src
    │   └── test
    └── wazuh_qa_framework
        ├── src
        │   ├── api_client
        │   ├── constants
        │   └── utils
        └── tests
            ├── api_client
            ├── constants
            └── utils
```

> tests/test_functional and tests/test_no_functional could be simplified by removing test prefix

I would prefer to keep the _"test_"_ prefix, as it helps _pytest_ to identify these directories. 
This can be useful if we want to run a specific selection of tests, some functional and some non-functional altogether, by only marking them with the same tag. Something like:
```bash
pytest  -m <tag>
```

**Comment by damarisg** [comment]
GJ!
